### PR TITLE
Import: Disable Cancel button while importing

### PR DESCRIPTION
**Pull request by vindl**
Hide Stop button during import process to prevent users from causing backfill failures when they needlessly click on it.

**Visual changes**

Before:
old

After:
new

### CODE REVIEW

**Comment by dllh** [comment]
Gave this a quick test. I canceled the import, and that worked as expected. Then I started again and started the import and did the user mapping. A "Start" screen showed for a few seconds before the progress bar displayed. When I checked the console, I found this:

screen shot 2017-01-19 at 9 32 16 am

The "Stop" button was gone as expected, and the import seems to be progressing as expected too. The code looks fine to me, though I'd defer to someone with more Calypso coding experience than I have for a :ship:.

**Comment by vindl** [comment]
@dllh The warning seems unrelated to this PR - tested it and it's occurring in `master` too. I tried to track it down quickly but no luck so far - I guess I'll file a separate issue for this.

**Comment by dmsnell** [comment]
Is this Medium-specific? We made the intentional decision with the WordPress importer to allow partial imports even thought it could result in messed-up state.

**Comment by vindl** [comment]
@dmsnell no, this affects all importers. I made it as a side PR while testing the Medium importer, according to comments from [this conversation](https://happinessgardeningp2.wordpress.com/2016/12/29/medium-importer-status-prompted-by/#comment-15439).

Just curious - what were the reasons for allowing partial imports?

**Comment by dmsnell** [comment]
> what were the reasons for allowing partial imports?

I can't remember exactly. it was something I raised as well. there are certainly cases where someone might accidentally start an import and then realize they made a mistake and we want to give them the ability to cancel before it gets too far.

the import process is unfortunately non-undoable so once you get going pieces come in one by one. it's not atomic in other words.

**Comment by dllh** [comment]
I don't remember the rationale for allowing partial imports. I guess maybe it had to do with the fact that we can't really roll back? At any rate, having lived with it for a while and suspecting that a lot of import backfill failures occur because the import is stopped early, I'm very much in favor of removing the stop button.

File: client/my-sites/importer/importer-header.jsx [hunk 1]
```diff
@@ -92,20 +92,27 @@ export const ImporterHeader = React.createClass( {
 		const canCancel = isEnabled && ! includes( [ appStates.UPLOADING ], importerState );
 		const isScary = includes( [ ...stopStates, ...cancelStates ], importerState );
 
+		// Users were clicking on Stop during artificial delay, which results in backfill failures.
+		// Temporarily introducing this in order to hide Stop button until async job is completed.
```
Comment (dmsnell) [inline client/my-sites/importer/importer-header.jsx]: this isn't temporary in any sense of the word. let's not kid ourselves :wink:

File: client/my-sites/importer/importer-header.jsx [hunk 2]
```diff
@@ -92,20 +92,27 @@ export const ImporterHeader = React.createClass( {
...
+						className="importer__master-control"
+						disabled={ ! canCancel }
+						isPrimary={ false }
+						scary={ isScary }
+						onClick={ this.controlButtonClicked }
+					>
+
+						{ this.getButtonText() }
+					</Button>
+				}
```
Comment (dmsnell) [inline client/my-sites/importer/importer-header.jsx]: Do we really want to hide the button? Notice that we already have `canCancel` controlling the disabled state. Maybe instead we should simply edit the expression above which determines `canCancel` and augment it with the cancelable state values.

File: client/my-sites/importer/importer-header.jsx (hunk 1 above)
Comment (vindl) [inline client/my-sites/importer/importer-header.jsx]: Updated. I don't believe in permanent code, but I see your point. :smile:

Comment (dmsnell) [inline client/my-sites/importer/importer-header.jsx]: believe it or not :wink:

besides, I didn't say it's permanent, but just that "this is temporary" is usually a lie and rarely defined. I mean, even the Roman empire was temporary

File: client/my-sites/importer/importer-header.jsx (hunk 2 above)
Comment (vindl) [inline client/my-sites/importer/importer-header.jsx]: It seems to be less confusing to me if we just hide it. With deactivated button, users might interpret that as stalled process, or wait for it to become active. In my opinion, hiding it communicates more clearly that they don't need to perform any more actions on that page, until they receive a notification.

Comment (dmsnell) [inline client/my-sites/importer/importer-header.jsx]: Let's double-check with a design review

**Comment by drw158** [comment]
What if the user wants to cancel the import? Seem like they should be able to do that. What are backfill errors?

**Comment by folletto** [comment]
I've read this and the other thread, and like Dave I'm not sure exactly what problem this is solving, and why it's solved by removing "Cancel" (which seems a good button to have).

It seems to be related to the backfilling of images, but what's the problem _exactly_?

**Comment by vindl** [comment]
@drw158 @folletto: The PR was inspired by Daryl's comment on that thread. Having said that, I might not be getting all the details of the problem right, but I'll try to explain it to the best of my knowledge. @dllh please correct me if I'm wrong, or provide a more detailed explanation. 

> I'm not sure exactly what problem this is solving

I think that the root of the problem is that we have an (arbitrary) artificial delay of 15 minutes built into the import process. If I understood correctly, this was introduced to allow for backfilling (downloading media that's linked in the export file) to complete before sending the notification that the import has finished. That means that even the smallest imports of just one post will take at least 15 minutes. The working assumption is that in some cases this is not matching user's expectations, and having no indicator of progress, they assume that the process has stalled and needlessly click on `Stop` (maybe in attempt to restart it). This causes the site to end up in a semi-imported state. 

> and why it's solved by removing "Cancel" (which seems a good button to have).

Removing the button can prevent cases like these, and allow imports to complete successfully. I'm aware that this solution is not ideal. Ideally we would have no artificial delay, meaningful progress bar that accurately keeps track of import progress, and `Cancel` functionality that roll-backs the data. However, implementing that is non-trivial and we probably won't be doing it anytime soon. So I think it boils down to deciding whether the proposed change is an improvement over the existing state or not. 

I agree that it makes sense to have `Cancel` button here, which would allow users to change their mind halfway through the process. On the other hand, I'm not sure how likely that scenario is after uploading the file and clicking on `Start Import` three times in a row (topic for another discussion/PR :smile:), or how valuable that action is considering that it is not rolling back the data that was imported up to that point.

**Comment by dllh** [comment]
@vindl you've stated the situation very clearly. In short, having the Stop button is causing people to wind up with broken sites because the backfill (the process by which we go through site content and update references to images within posts to point to the newly imported image urls) never happens. Further, this is a silent failure; the user never knows it's an issue until they cancel their old hosting or point their DNS to us, at which point the images (which still point to their old locations) are suddenly broken and sometimes irretrievably so. Removing the Stop button has no such negative effects.

**Comment by folletto** [comment]
Thank you! That's a great clarification. :)

I know there are more/different internal states, but in terms of the user interactions, would I be correct in assuming that the following details the ideal various states of the UI:

1. Before Starting — idle, show "Start Import" button
2. Waiting File — show the drop area, show "Cancel" button (can we rename "Close"?)
3. Importing — the data from the file is being imported, show "Stop" button
4. Backfilling — the data has been imported from the file, but the images are back-filling, show NO button
5. Done — the transfer is complete, show message, notify user

Is this correct?

**Comment by vindl** [comment]
@folletto I believe that's pretty much it, just one minor detail to add - between steps 2 and 3, after the export archive is uploaded, and before the actual import starts, there is an intermediate step for defining user mapping which specifies how will the imported posts be assigned to existing users of that blog. 
(step 4 is currently not exposed and it sounds like implementation detail, not sure if there is value in showing it or how hard that would be)

**Comment by folletto** [comment]
> (step 4 is currently not exposed and it sounds like implementation detail, not sure if there is value in showing it or how hard that would be)

Ok!

If we can, then it would partially "solve" the ambiguity above: Stop will still show when the import is partial (with the old "partially imported" message), at the same time it allows it to not break in the backfilling (with the new "notify" message).

Another detail: if we already know that it takes 15 minutes, it's actually AWESOME as it's a clear margin: let's say it! :)

**Comment by drw158** [comment]
Thank you very much for the explanation. Because it's non-trivial to get the ideal solution working, I think we just need to make the best of it. Sounds like we really need to disable the stopping. I'd like some feedback from @folletto but I think this might be better:

- Don't hide the button, just disable it and change the button text to `Importing..."
- Perhaps move the spinner adjacent to the button so we can make sure the user sees it (see image below)
- Because the spinner has been moved we can left-align the explanation text.
- If the import will always take at least 15 minutes, maybe we can tweak the text to set that expectation.

> Importing will take some time, but you can safely...


screen shot 2017-01-20 at 12 43 24 pm

**Comment by dmsnell** [comment]
> 3. Importing — the data from the file is being imported, show "Stop" button
> 4. Backfilling — the data has been imported from the file, but the images are back-filling, show NO button

if I did my homework properly way back when, these two aren't sequential processes. for each post we fire off an asynchronous job to backfill that particular post. this means that the backfilling starts with the first import and may extend beyond the last post import. they are simultaneous.

p2XLIi-1gl (hope this link is right - should be a diagram of the import process)

**Comment by folletto** [comment]
> I think this might be better

Seems a good solution to me! :)


> if I did my homework properly way back when, these two aren't sequential processes. for each post we fire off an asynchronous job to backfill that particular post. this means that the backfilling starts with the first import and may extend beyond the last post import. they are simultaneous.

Efficient! Ok, so we don't have really two states there. 

So I'd say: let's go with Dave's proposal above, and let's be super clear it is going to take no less than 15 minutes  ;)

Here's a copy idea:

> "Importing should take a bit more than 15 minutes, sometimes more. You can safely navigate away from this page if you need to. "

File: client/my-sites/importer/importer-header.jsx [hunk 3]
```diff
@@ -71,15 +71,15 @@ export const ImporterHeader = React.createClass( {
...
 		}
 
 		if ( includes( cancelStates, importerState ) ) {
-			return this.translate( 'Cancel', { context: 'verb' } );
+			return this.translate( 'Close', { context: 'verb' } );
 		}
 
 		if ( includes( stopStates, importerState ) ) {
-			return this.translate( 'Stop Import', { context: 'verb' } );
+			return this.translate( 'Importing…', { context: 'verb' } );
```
Comment (a8ci18n) [inline client/my-sites/importer/importer-header.jsx]: Alternate string suggestion: 
* `translate( 'Importing...' )`&mdash; translations: **71**. *ES Score: 11.44*

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/remove-stop-during-import)

File: client/my-sites/importer/importer-header.jsx [hunk 4]
```diff
@@ -71,15 +71,15 @@ export const ImporterHeader = React.createClass( {
 		const { importerState } = this.props.importerStatus;
 
 		if ( includes( startStates, importerState ) ) {
-			return this.translate( 'Start Import', { context: 'verb' } );
+			return this.translate( 'Start', { context: 'verb' } );
 		}
 
 		if ( includes( cancelStates, importerState ) ) {
-			return this.translate( 'Cancel', { context: 'verb' } );
+			return this.translate( 'Close', { context: 'verb' } );
```
Comment (a8ci18n) [inline client/my-sites/importer/importer-header.jsx]: Alternate string suggestion: 
* `translate( 'Close', { context: 'verb, to Close a dialog'} )`&mdash; translations: **23**. *ES Score: 6.60*

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/remove-stop-during-import)

File: client/my-sites/importer/importer-header.jsx [hunk 5]
```diff
@@ -71,15 +71,15 @@ export const ImporterHeader = React.createClass( {
 		const { importerState } = this.props.importerStatus;
 
 		if ( includes( startStates, importerState ) ) {
-			return this.translate( 'Start Import', { context: 'verb' } );
+			return this.translate( 'Start', { context: 'verb' } );
```
Comment (a8ci18n) [inline client/my-sites/importer/importer-header.jsx]: Warning: 30 translations will be lost with this change.
Alternate string suggestion: 
* `translate( 'Start' )`&mdash; translations: **31**. *ES Score: 5.14*

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/remove-stop-during-import)

File: client/my-sites/importer/importer-header.jsx [hunk 6]
```diff
@@ -75,11 +75,11 @@ export const ImporterHeader = React.createClass( {
 		}
 
 		if ( includes( cancelStates, importerState ) ) {
-			return this.translate( 'Cancel', { context: 'verb' } );
+			return this.translate( 'Close', { context: 'verb, to Close a dialog' } );
 		}
 
 		if ( includes( stopStates, importerState ) ) {
-			return this.translate( 'Stop Import', { context: 'verb' } );
+			return this.translate( 'Importing...', { context: 'verb' } );
```
Comment (a8ci18n) [inline client/my-sites/importer/importer-header.jsx]: Alternate string suggestion: 
* `translate( 'Importing...' )`&mdash; translations: **71**. *ES Score: 11.44*

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/remove-stop-during-import)

**Comment by vindl** [comment]
@drw158 I think this is ready for another go: 
- Updated button to disabled state and text to `Importing...`. Also changed button style to `primary` since it was hardly visible with default styles when disabled. 
- I didn't move the spinner since it the progress bar appears there shortly (updated PR description with screenshots).
- Updated the text as suggested.
- Changed `Cancel` button label to `Close`. 
- Changed `Start Import` button label in intermediate state to `Continue`, to avoid repeating it multiple times.

**Comment by drw158** [comment]
Just checked it. Design looks good 👍

File: client/my-sites/importer/importing-pane.jsx [hunk 7]
```diff
@@ -115,15 +115,8 @@ export const ImportingPane = React.createClass( {
...
-			'Importing may take a while, but you can ' +
-			'safely navigate away from this page if you need ' +
-			'to. If you {{b}}stop the import{{/b}}, your site ' +
-			'will be {{b2}}partially imported{{/b2}}.', {
-				components: {
-					b: <strong />,
-					b2: <strong />
-				}
-			}
+			'Importing should take 15 minutes, sometimes more if your site has a lot of media. ' +
```
Comment (roundhill) [inline client/my-sites/importer/importing-pane.jsx]: Maybe add an `about` in there so they don't expect it to be exactly 15 minutes?

`Importing should take about 15 minutes...`

Comment (folletto) [inline client/my-sites/importer/importing-pane.jsx]: Doesn't the "sometimes more" bit cover that? :D

Comment (roundhill) [inline client/my-sites/importer/importing-pane.jsx]: Oh! I suppose so. Could it ever be shorter than 15 minutes?

Comment (folletto) [inline client/my-sites/importer/importing-pane.jsx]: From the discussion above, no:

> we have an (arbitrary) artificial delay of 15 minutes built into the import process

Comment (vindl) [inline client/my-sites/importer/importing-pane.jsx]: > From the discussion above, no:

That's correct, it will take at least 15 minutes.

Comment (roundhill) [inline client/my-sites/importer/importing-pane.jsx]: OK! What do you think of:

`Importing takes 15 minutes, or a while longer if your site has a lot of media.`

Comment (dmsnell) [inline client/my-sites/importer/importing-pane.jsx]: @roundhill that comma doesn't belong there. How about something like this…

> The import process requires at least fifteen minutes to complete but may take longer depending on how many media items are in the site.

…maybe we can get some editorial review?

Comment (roundhill) [inline client/my-sites/importer/importing-pane.jsx]: Your sentence is really long :) 

+1 for editorial.

Comment (roundhill) [inline client/my-sites/importer/importing-pane.jsx]: We can always change this later, too. The current copy gets the message across.

Comment (drw158) [inline client/my-sites/importer/importing-pane.jsx]: How about this 

`Importing takes 15 minutes or a while longer if your site has a lot of media.`

Short, conversation, and there is no comma :)

File: client/my-sites/importer/importer-header.jsx (hunk 6 above)
Comment (a8ci18n) [inline client/my-sites/importer/importer-header.jsx]: Hi! I've found a possible matching string that has already been translated **71** times:
`translate( 'Importing...' )` *ES Score: 6.69*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/importer/importing-pane.jsx (hunk 7 above)
Comment (vindl) [inline client/my-sites/importer/importing-pane.jsx]: Updated according to @drw158's suggestion. Thanks for help everyone!

File: client/my-sites/importer/importer-header.jsx (hunk 6 above)
Comment (yoavf) [inline client/my-sites/importer/importer-header.jsx]: @vindl implementing this suggestion would have been good. Adding a context just adds a new string that will have to be translated again.

Comment (vindl) [inline client/my-sites/importer/importer-header.jsx]: @yoavf if it's not too late I'll spin up a PR to remove the context.

Comment (yoavf) [inline client/my-sites/importer/importer-header.jsx]: No problem, it's never too late to consolidate strings ☺️

**Pull request merged**